sql/core/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -757,7 +758,10 @@ private[sql] class SQLConf extends Serializable with CatalystConf with Logging {
 
   def variableSubstituteDepth: Int = getConf(VARIABLE_SUBSTITUTE_DEPTH)
 
-  def warehousePath: String = new Path(getConf(WAREHOUSE_PATH)).toString
+  def warehousePath: String = {
+    val path = new Path(getConf(WAREHOUSE_PATH))
+    FileSystem.get(path.toUri, new Configuration()).makeQualified(path).toString
```

Comment:
URI standard might be too restrictive for something that a user has to manually specify as a string. According to [RFC 3986](https://tools.ietf.org/html/rfc3986#section-2.3), it allows only numbers, digits, dash, underscore, dot and tilde. Anything else is required to be [percent-encoded](https://tools.ietf.org/html/rfc3986#section-2.1). This means that we have to require user to always do percent-encoding of her path if we stick with URI. There are other paths in Spark that user can specify and they are not required to be encoded. This might add unnecessary confusion. I would suggest to keep it as a path, document that one can optionally add a schema if the path is not local, and try to handle different options within Spark.